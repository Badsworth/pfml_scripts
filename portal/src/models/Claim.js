/* eslint sort-keys: ["error", "asc"] */
/**
 * @file Benefits application model and enum values
 */
import BaseModel from "./BaseModel";

class Claim extends BaseModel {
  get defaults() {
    return {
      application_id: null,
      // TODO: We'll map this to the correct API field once we get into Intermittent Leave work
      avg_weekly_hours_worked: null,
      created_at: null,
      duration_type: null,
      employee_ssn: null,
      employer_benefits: [],
      employer_fein: null,
      // TODO: This field does not yet exist in the API: https://lwd.atlassian.net/browse/CP-574
      employment_status: null,
      first_name: null,
      // TODO: We'll map this to the correct API field once we get into Intermittent Leave work
      hours_off_needed: null,
      last_name: null,
      leave_details: {
        continuous_leave_periods: null,
        employer_notification_date: null,
        employer_notified: null,
        reason: null,
      },
      middle_name: null,
      other_incomes: [],
      pregnant_or_recent_birth: null,
      previous_leaves: [],
      status: null,
    };
  }
}

/**
 * Enums for the Application's `status` field
 * @enum {string}
 */
export const ClaimStatus = {
  // Submitted by the user (Step 2)
  completed: "Completed",
  // AKA: In Progress (Step 1)
  started: "Started",
  // Stored in the claims processing system (Step 3)
  submitted: "Submitted",
};

/**
 * Enums for the Application's `leave_details.employment_status` field
 * @enum {string}
 */
export const EmploymentStatus = {
  employed: "employed",
  selfEmployed: "self-employed",
  unemployed: "unemployed",
};

/**
 * Enums for the Application's `leave_details.reason` field
 * @enum {string}
 */
export const LeaveReason = {
  // TODO: We need to map some of these to the correct API fields,
  // once those enum values exist within the API
  // https://lwd.atlassian.net/browse/CP-515
  activeDutyFamily: "Care For A Family Member",
  bonding: "Child Bonding",
  medical: "Serious Health Condition - Employee",
  serviceMemberFamily: "Pregnancy/Maternity",
};

export default Claim;
