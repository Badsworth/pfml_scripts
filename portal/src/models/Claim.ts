/* eslint sort-keys: ["error", "asc"] */
import { compact, get } from "lodash";

/**
 * A record from the API's Claims table. Could be utilized by Leave Admin and Claimants.
 */
class Claim {
  absence_period_end_date: string | null = null;
  absence_period_start_date: string | null = null;
  claim_status: AbsenceCaseStatusType | null;
  claim_type_description: string | null = null;
  created_at: string;
  employee: ClaimEmployee | null;
  employer: ClaimEmployer;
  fineos_absence_id: string;
  fineos_notification_id: string;
  managed_requirements: ManagedRequirement[];

  constructor(attrs: Claim) {
    Object.assign(this, attrs);
    if (attrs.employee) {
      this.employee = new ClaimEmployee(attrs.employee);
    }
  }
}

/**
 * Employee (claimant) record associated to the Claim
 */
export class ClaimEmployee {
  first_name: string | null = null;
  last_name: string | null = null;
  middle_name: string | null = null;
  other_name: string | null = null;

  constructor(attrs: Partial<ClaimEmployee>) {
    Object.assign(this, attrs);
  }

  get fullName() {
    return compact([
      get(this, "first_name"),
      get(this, "middle_name"),
      get(this, "last_name"),
    ]).join(" ");
  }
}

/**
 * Employer record associated to the Claim
 */
export interface ClaimEmployer {
  employer_dba: string | null;
  employer_fein: string;
  employer_id: string;
}

/**
 * Managed requirements associated to the Claim
 */
export interface ManagedRequirement {
  category: string;
  created_at: string;
  follow_up_date: string | null;
  responded_at: string | null;
  status: "Open" | "Complete" | "Suppressed";
  type: string;
}

/**
 * Enums for `claim.claim_status` field
 * This is the subset of potential values returned by the API that we show to the user.
 * We currently ignore other potential values like "Adjudication", "Intake In Progress", "Unknown"
 * @enum {string}
 */
export const AbsenceCaseStatus = {
  approved: "Approved",
  closed: "Closed",
  completed: "Completed",
  declined: "Declined",
} as const;

export type AbsenceCaseStatusType =
  typeof AbsenceCaseStatus[keyof typeof AbsenceCaseStatus];

export default Claim;
