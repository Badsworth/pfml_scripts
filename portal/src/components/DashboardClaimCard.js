import ButtonLink from "../components/ButtonLink";
import Claim from "../models/Claim";
import Heading from "../components/Heading";
import PropTypes from "prop-types";
import React from "react";
import routeWithParams from "../utils/routeWithParams";
import { useTranslation } from "../locales/i18n";

/**
 * Item displayed in the "Active claims" section of the dashboard, for
 * an individual in progress claim
 */
function DashboardClaimCard(props) {
  const { claim, number } = props;
  const { t } = useTranslation();

  return (
    <article className="maxw-mobile-lg border border-base-lighter margin-bottom-3 padding-3">
      <span className="usa-tag bg-primary-lighter display-inline-block margin-bottom-1 text-bold text-primary text-no-uppercase">
        {t("pages.index.tagInProgress")}
      </span>
      <Heading className="margin-top-05 margin-bottom-2" level="3">
        {t("pages.index.claimCardHeading", { number })}
      </Heading>

      <ButtonLink
        className="margin-top-1"
        href={routeWithParams("claims.checklist", {
          claim_id: claim.application_id,
        })}
      >
        {t("pages.index.resumeClaimButton")}
      </ButtonLink>
    </article>
  );
}

DashboardClaimCard.propTypes = {
  claim: PropTypes.instanceOf(Claim),
  /**
   * Cards are displayed in a list. What position is this card?
   */
  number: PropTypes.number.isRequired,
};

export default DashboardClaimCard;
