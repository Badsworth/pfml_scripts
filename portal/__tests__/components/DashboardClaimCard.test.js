import Claim from "../../src/models/Claim";
import DashboardClaimCard from "../../src/components/DashboardClaimCard";
import React from "react";
import { shallow } from "enzyme";

describe("DashboardClaimCard", () => {
  it("renders a card for the given claim", () => {
    const wrapper = shallow(
      <DashboardClaimCard
        claim={new Claim({ application_id: "mock-claim-id" })}
        number={2}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
