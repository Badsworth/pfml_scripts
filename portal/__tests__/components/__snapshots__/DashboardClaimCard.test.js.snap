// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardClaimCard renders a card for the given claim 1`] = `
<article
  className="maxw-mobile-lg border border-base-lighter margin-bottom-3 padding-3"
>
  <span
    className="usa-tag bg-primary-lighter display-inline-block margin-bottom-1 text-bold text-primary text-no-uppercase"
  >
    In Progress
  </span>
  <Heading
    className="margin-top-05 margin-bottom-2"
    level="3"
  >
    Claim 2
  </Heading>
  <ButtonLink
    className="margin-top-1"
    href="/claims/name?claim_id=mock-claim-id"
  >
    Finish your claim
  </ButtonLink>
</article>
`;
